In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score
from sklearn import linear_model, tree, ensemble

In [2]:
td = pd.read_csv('petrol_consumption2.csv')

In [3]:
td.head()

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [4]:
y = td.Petrol_Consumption          
X = td.drop(['Petrol_Consumption'], axis=1)

In [9]:
kf = KFold(n_splits=5, shuffle=True, random_state=42)
c = 1
for train_index, test_index in kf.split(X):
    print(f'Train {c} :', train_index, 'Test:', test_index)
    c += 1

Train 1 : [ 0  1  2  3  5  6  7  8  9 10 11 13 14 15 16 17 18 20 21 22 23 28 29 30
 31 32 33 34 35 36 38 39 41 42 44 45 46 47] Test: [ 4 12 19 24 25 26 27 37 40 43]
Train 2 : [ 0  1  2  4  5  7 10 11 12 14 16 18 19 20 21 22 23 24 25 26 27 28 29 30
 31 32 34 35 36 37 38 40 41 42 43 44 46 47] Test: [ 3  6  8  9 13 15 17 33 39 45]
Train 3 : [ 1  2  3  4  6  7  8  9 10 12 13 14 15 17 18 19 20 21 22 23 24 25 26 27
 28 33 35 36 37 38 39 40 41 42 43 44 45 47] Test: [ 0  5 11 16 29 30 31 32 34 46]
Train 4 : [ 0  3  4  5  6  7  8  9 11 12 13 14 15 16 17 18 19 20 22 24 25 26 27 28
 29 30 31 32 33 34 37 38 39 40 42 43 45 46 47] Test: [ 1  2 10 21 23 35 36 41 44]
Train 5 : [ 0  1  2  3  4  5  6  8  9 10 11 12 13 15 16 17 19 21 23 24 25 26 27 29
 30 31 32 33 34 35 36 37 39 40 41 43 44 45 46] Test: [ 7 14 18 20 22 28 38 42 47]


In [10]:
score = cross_val_score(linear_model.LinearRegression(), X, y, cv= kf)
print(f'Scores for each fold: {score}')
np.sqrt(score.mean())

Scores for each fold: [0.3913664  0.63392548 0.05222032 0.41270325 0.76147919]


0.671072968904877

**Використав к-кратну кросвалідацію він здався мені найбліш зрозумілим, по даним оцінки модель не можна використовувати, так як оцінка моделі вийшла досить слабкою (0.67).**